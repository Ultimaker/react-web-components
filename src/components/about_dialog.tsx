import * as React from 'react';
import { I18n } from '../utils/i18n';

import Button from './button';
import Modal from "./modal";
import Tile from "./tile";


/**
 * The translated messages for this component.
 */
const T = {
  // translate beforehand
  _aboutApp: I18n.translate("About dialog", "About %{appName}"),
  _license: I18n.translate("About dialog", "license %{license}"),
  _version: I18n.translate("About dialog", "Version: %{versionNumber}"),
  _appUses: I18n.translate("About dialog", "%{appName} uses the following Open Source components:"),

  // only interpolate when needed
  aboutApp: (appName: string) => I18n.interpolate(T._aboutApp, {appName}),
  license: (license: string) => I18n.interpolate(T._license, {license}),
  version: (versionNumber: any) => I18n.interpolate(T._version, {versionNumber}),
  appUses: (appName: string) => I18n.interpolate(T._appUses, {appName}),

  forSupport: I18n.translate("About dialog", "For support visit:"),
}


export interface AboutDialogProps {
  /** List of open source components used */
  componentsList: OpenSourceComponent[];
  /** Name of the application */
  appName: string;
  /** Version number of the application */
  versionNumber: string;
  /** Dialog will be displayed when true */
  isOpen: boolean;
  /** Function to close the dialog */
  closeHandler: () => void;
  /** URL to a support page */
  supportLinkURL?: string;
  /** Link text for the support page */
  supportLinkText?: string;
}

interface OpenSourceComponent {
  name: string;
  license: string;
  url: string;
}


function licenseList(componentsList: OpenSourceComponent[]): JSX.Element[] {
  return componentsList.map(component => (
      <tr key={component.name}>
        <td className="about-component-name">
          <a href={component.url} target="_blank">{component.name}</a>
        </td>
        <td className="about-component-license">
          <a href={"https://spdx.org/licenses/" + component.license + ".html"} target="_blank">
            {T.license(component.license)}
          </a>
        </td>
      </tr>
    )
  );
}

const AboutDialog: React.StatelessComponent<AboutDialogProps> =
  ({ componentsList, isOpen, versionNumber, closeHandler, appName, supportLinkURL, supportLinkText }) => {

    return (
      <div className="about-dialog">
        <Modal isOpen={isOpen} onOverlayClickHandler={closeHandler}>
          <Tile>
            <div className="about-dialog__header">{T.aboutApp(appName)}</div>
            <p>{T.version(versionNumber)}</p>
            {supportLinkURL &&
              <p>{T.forSupport}&nbsp;&nbsp;&nbsp;<a href={supportLinkURL} target="_blank">{supportLinkText}</a></p>
            }
            <p>{T.appUses(appName)}</p>
            <table className="about-components-list">
              <tbody>
                {licenseList(componentsList)}
              </tbody>
            </table>
            <div className="about-dialog__btn">
              <Button onClickHandler={closeHandler}>Close</Button>
            </div>
          </Tile>
        </Modal>
      </div>
    );
  };

AboutDialog.displayName = "AboutDialog";

export default AboutDialog;
