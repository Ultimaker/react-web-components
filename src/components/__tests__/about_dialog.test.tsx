// Copyright (c) 2018 Ultimaker B.V.
import * as React from 'react';
import { shallow } from 'enzyme';

// component
import { AboutDialog, AboutDialogProps } from '../about_dialog';

describe('The AboutDialog component', () => {
    let props: AboutDialogProps;
    let wrapper;

    beforeEach(() => {
        props = {
            closeHandler: jest.fn(),
            componentsList: [{
                name: 'testComponent',
                license: 'MIT',
                url: 'https://ultimaker.com/',
            }],
            versionNumber: '1.0.0',
            headerText: 'About',
            packagesPreText: 'This app uses the following Open Source components:',
            licensePreText: 'license',
            versionPreText: 'Version',
        };
        wrapper = shallow(<AboutDialog {...props} />);
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
        expect(props.closeHandler).not.toHaveBeenCalled();
    });
});
