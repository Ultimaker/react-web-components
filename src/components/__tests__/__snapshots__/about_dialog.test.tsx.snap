// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The AboutDialog component should render 1`] = `
<div
  className="about-dialog"
>
  <PopupBase
    headerText="About test app"
    width="md"
  >
    <p>
      Version: 1.0.0
    </p>
    <p>
      For support visit:
         
      <a
        href="https://ultimaker.support.com/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Support link
      </a>
    </p>
    <p>
      test app uses the following Open Source components:
    </p>
    <table
      className="about-components-list"
    >
      <tbody>
        <tr
          key="testComponent"
        >
          <td
            className="about-component-name"
          >
            <a
              href="https://ultimaker.com/"
              rel="noopener noreferrer"
              target="_blank"
            >
              testComponent
            </a>
          </td>
          <td
            className="about-component-license"
          >
            <a
              href="https://spdx.org/licenses/MIT.html"
              rel="noopener noreferrer"
              target="_blank"
            >
              license MIT
            </a>
          </td>
        </tr>
      </tbody>
    </table>
    <FormActions
      align="right"
    >
      <Button
        appearance="primary"
        className=""
        linkToNewTab={false}
        onClickHandler={[MockFunction]}
        shape="rectangle"
        type="button"
      >
        Close
      </Button>
    </FormActions>
  </PopupBase>
</div>
`;
